microsoft/iot-cardboard-js
# Cardboard lib src structure V2 📂

# Change proposal
https://www.notion.so/iot-cardboard-js/Cardboard-lib-src-structure-V2-32b0ff2aa3bc43919393f0e673af2f4d
This document is in a notion workspace I created for our team. You'll need to create a notion account with your @microsoft email to access.
# Motivation

The **cardboard** component library has shifted away from the concept of **Cards**, but in many places we’re still using **`Cards`** and **`CompositeCards`** to structure the library.  This refactor will help simplify, add consistency, and add clarity to the library structure.

Along with this refactor (in another PR) — The node auto templating script can be updated to make new component creation easier.

![image](https://user-images.githubusercontent.com/18181049/154756107-19e1ed06-d276-4e48-adca-24f605802e5b.png)

# Proposed **changes**

### **Pages**

Top level pages will live here.  This includes anything that is considered a *page* from an application point of view.  Currently *pages* include the */flights* pages:

- Introduce `Pages/` for top level application page components
- **`ADT3DScenePage`** aka ‘3D’
- ~~ADTModelListWithModelDetailsCard~~ **`ADTModelAuthoringPage`** aka ‘Models’
- ~~ADTHierarchyWithBoard~~ **`ADTTwinsPage`** aka 'Twins'
- Update rollup library build to generate direct imports for `Pages`

![image](https://user-images.githubusercontent.com/18181049/154756127-b3f79c7d-a187-4c6e-9ae0-07f0e0e60380.png)

### Cards

Cards are React components which use the `BaseCard` component to wrap their rendered element.  This will ensure that error handling, loading states, theming, and localization is consistent between all cards (`BaseCard` wraps `BaseComponent`  in additional styles, and add a title header). Cards can be used by both Boards and in a standalone context. 

- Remove`Consume` and `Create`separation for Cards
- Refactor select components out of `CompositeCards`
    - `ADT3DSceneBuilder` → `Components/` (not using `CompositeCard`)
    - `ADT3DScenePage` → `Pages/ADT3DScenePage` (Top level page)
    - `ADTHierarchyWithBoard` → `Pages/ADTTwinPage` (Top level *page)*
    - `ADTModelListWithModelDetails` → `Pages/ADTModelAuthoringPage` (Top level *page)*
    - `BimUpload` → `Cards/BimUploadCard` (Using `BaseCard` not `CompositeCard`)
        
![image](https://user-images.githubusercontent.com/18181049/154756141-b8ff9e64-bb49-4f20-8150-5c4adb71d269.png)

### Components

**iot-cardboard-js** React components live here.  Components will have Storybook files associated with them for snapshot / interaction testing and isolated development.  A complex component may compose ****other components and have nested **Internal** components.  These will all live inside that components root folder.

- Move components incorrectly created as ‘Cards’ i.e. no `BaseCard` usage into `Components/`
- Group React components used **only** internally to a component into an `Internal/` folder

![image](https://user-images.githubusercontent.com/18181049/154756161-80712c8e-4aa9-4f1f-a37c-267a8cfbf50e.png)
        
### Models

- Move `Models/Utilities/StoryUtilities` → `Models/Services/StoryUtilities`

### Stories

- Update story files for organized viewing & to reflect structure changes

![image](https://user-images.githubusercontent.com/18181049/154756937-6e85e8b1-94c4-495c-814d-c8f9112fefd8.png)


### Testing 🧪
- Verify each story runs as expected
- Verify build runs as expected
- Verify folder structure makes sense to you

### Checklist ✔️
- [x] Added to Cardboard kanban project
- [x] Requested reviews
